modules/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageService.java

```diff
@@ -140,36 +165,34 @@ private Storage createClient(GoogleCloudStorageClientSettings gcsClientSettings,
             try (InputStream keyStoreStream = GoogleUtils.class.getResourceAsStream("google.jks")) {
                 SecurityUtils.loadKeyStore(certTrustStore, keyStoreStream, "notasecret");
             }
-            builder.trustCertificates(certTrustStore);
-            Proxy proxy = gcsClientSettings.getProxy();
-            if (proxy != null) {
-                builder.setProxy(proxy);
-                notifyProxyIsSet(proxy);
-            }
-            return builder.build();
-        });
+            final SSLContext sslContext = SSLContexts.custom().loadTrustMaterial(certTrustStore, null).build();
+            final GoogleCloudStorageHttpStatsCollector httpStatsCollector = new GoogleCloudStorageHttpStatsCollector(stats);
+            final HttpClientBuilder builder = HttpClients.custom()
+                .addInterceptorLast(httpStatsCollector)
+                .setSSLContext(sslContext)
+                .setConnectionManager(connectionManager.updateAndGet(existing -> {
+                    if (existing == null) {
+                        final PoolingHttpClientConnectionManager connectionManager = new PoolingHttpClientConnectionManager();
+                        connectionManager.setMaxTotal(maxOpenConnections);
+                        connectionManager.setDefaultMaxPerRoute(maxOpenConnections);
+                        return connectionManager;
```

Comment:
There is no setting to purge idle connections from the pool (see "Connection Eviction Policy" [here](https://hc.apache.org/httpcomponents-client-4.5.x/current/tutorial/html/connmgmt.html)).

We would have to schedule a task to do that.

Reply:
`org.apache.http.impl.client.HttpClientBuilder#evictExpiredConnections` doesn't work if we want to share a thread pool.

> Please note this method has no effect if the instance of HttpClient is configuted[sic] to use a shared connection manager.